python-sdk/nuscenes/predict/helper.py

```diff
@@ -0,0 +1,235 @@
+# nuScenes dev-kit.
+# Code written by Freddy Boulton, 2020.
+
+from typing import Dict, Tuple, Any, List, Callable
+
+import numpy as np
+from pyquaternion import Quaternion
+
+from nuscenes import NuScenes
+from nuscenes.eval.common.utils import quaternion_yaw
+
+MICROSECONDS_PER_SECOND = 1e6
+BUFFER = 0.15  # seconds
+
+
+def angle_of_rotation(yaw: float) -> float:
+    """
+    Given a yaw angle (measured from x axis), find the angle needed to rotate by so that
+    the yaw points upwards (pi / 2).
+    :return: TODO.
+    """
+    return (np.pi / 2) + np.sign(-yaw) * np.abs(yaw)
+
+
+def convert_global_coords_to_local(coordinates: np.ndarray,
+                                   translation: Tuple[float, float, float],
+                                   rotation: Tuple[float, float, float, float]) -> np.ndarray:
+    """
+    Converts global coordinates to coordinates in the frame given by the rotation quaternion and
+    centered at the translation vector. The rotation is meant to be a z-axis rotation.
+    :param coordinates: x,y locations. array of shape [n_steps, 2]
+    :param translation: Tuple of (x, y, z) location that is the center of the new frame
+    :param rotation: Tuple representation of quaternion of new frame.
+        Representation - cos(theta / 2) + (xi + yi + zi)sin(theta / 2).
+    :return: TODO.
+    """
+    yaw = angle_of_rotation(quaternion_yaw(Quaternion(rotation)))
+
+    transform = np.array([[np.cos(yaw), -np.sin(yaw)], [np.sin(yaw), np.cos(yaw)]])
+
+    coords = (coordinates - np.atleast_2d(np.array(translation)[:2])).T
+
+    return np.dot(transform, coords).T[:, :2]
+
+
+class PredictHelper:
+    """ Wrapper class around NuScenes to help retrieve data for the prediction task. """
+
+    def __init__(self, nusc: NuScenes):
+        """
+        Inits PredictHelper
+        :param nusc: Instance of NuScenes class
+        """
+        self.data = nusc
+        self.inst_sample_to_ann = self._map_sample_and_instance_to_annotation()
+
+    def _map_sample_and_instance_to_annotation(self) -> Dict[Tuple[str, str], str]:
+        """
+        Creates mapping to look up an annotation given a sample and instance in constant time.
+        :return: TODO.
+        """
+        mapping = {}
+
+        for record in self.data.sample_annotation:
+            mapping[(record['sample_token'], record['instance_token'])] = record['token']
+
+        return mapping
+
+    def _timestamp_for_sample(self, sample: str) -> float:
+        """
+        Gets timestamp from sample token.
+        :param sample: TODO.
```

Comment:
I suggest to add `_token` to each token to make it clearer that this is not the sample dict.